### ISSUE TITLE
Message.__del__ can lead to file descriptor deadlock

### ISSUE THREAD

**Issue by joell** [ldo/dbussy#18]
_tl;dr: Relying on `Message.__del__` to call `dbus_message_unref()` can leave duplicate file descriptors open that can cause deadlock scenarios depending on what the Python garbage collector does.  Replacing reliance on `__del__` with explicit cleanup of Message objects in Ravel will prevent these deadlocks._

libdbus calls `dup()` both when a file descriptor is added to a request DBusMessage as an argument and when a file descriptor is retrieved from a reply DBusMessage as a result.  Since a DBussy Message wraps a libdbus DBusMessage, this means that every Message object is carrying around duplicates for any file descriptors added to or extracted from it.

DBussy uses the `Message.__del__` method to call libdbus' `dbus_message_unref()`, which closes those duplicate file descriptors.

The combination of these two facts can result in deadlock during scenarios where a DBus client or service is depending on the other side to close a file descriptor but the Python garbage collector hasn't reaped the `Message` object.  An example would be when a client sends the writer FD of a pipe to a DBus service as a means to receive a bulk data transfer.  Even if the service Python code closes the file descriptor request argument it receives from DBussy, its end of the pipe will remain open if the dangling Message object hasn't been picked up by the Python garbage collector and its `__del__` method invoked.

You can reproduce this as follows:

1. Run a small Ravel DBus service that exposes an interface that accepts one file descriptor, writes b"Hello" to it, closes it, and returns a success DBus response.
2. Run a small Ravel DBus client that:
    1. Creates a pipe via `r,w = os.pipe()`.
    2. Sends the `w` file descriptor to the DBus service as an request argument.
    3. Waits for a successful response from the DBus service.
    4. Attempts to read what the service sent, via `with open(r, "rb") as f: f.read()`
3. Observe that on some runs the client will block indefinitely waiting for more data from the service's end of the pipe.  (Sometimes it may not block, if the Python GC on the service side collects the Message object.  In those cases, just run the client again without restarting the service; within 2-3 requests I usually hit the scenario in the service code where the Python GC doesn't reap the Message object.)

This issue is most visible on the service side, but you can also run into it on the client side as well where a Message object isn't collected and its `dup()` of the file descriptor that was passed in a DBus request is still open.

This issue should be resolvable by replacing `Message.__del__` with a cleanup method that must be explicitly invoked when a Message is done with (and updating Ravel to use it to explicitly clean up all Message objects it creates on the user's behalf).

**Comment by ldo** [comment]
First of all, I cannot reproduce any such hang. 
[Enclosed](https://github.com/ldo/dbussy/files/3336885/fd_reap_test.zip) is the pair of sample scripts I wrote, trying to follow your description; let me know if I misunderstood anything. I can run the client for any number of request/response cycles (I tried various numbers up to 1000), and it works just fine.

Secondly, I have absolutely no intention of adding any such explicit cleanup calls. The whole point of a dynamic language like Python is precisely so it handles such resource-management headaches automatically. And the important point about CPython, in particular, is that it does reference-counting as a first resort before falling back to garbage collection, so that abandoned objects disappear sooner, rather than later, and you do not have to resort to the Java-standard tactic of chaining up your program inside a fixed-size heap.

**Comment by joell** [comment]
Thank you for such a quick response and for taking the time to try to reproduce this issue!  I greatly appreciate it!

I ran your sample scripts, and there was one small and very important detail from my reproduction steps that was missed: Your call to `os.read(ın, 512)` will fetch the `b"Hello"` data and then return _without_ attempting to reach EOF on the read pipe, in contrast to how `with open(ın, "rb") as f: f.read()` would have.  As such, `fd_reap_client` never tests if the server actually closed the write end of the pipe and therefore never reproduces the dealock.

When I added a second `os.read` call to `fd_reap_client`, as illustrated in the patch below, the client hung on the very first iteration.
```diff
--- a/fd_reap_client
+++ b/fd_reap_client
@@ -42,6 +42,9 @@ async def mainline() :
         data = os.read(ın, 512)
         sys.stdout.write("attempt: %d: got %s\n" % (i + 1, repr(data)))
         sys.stdout.flush()
+        try_to_reach_eof = os.read(ın, 512)
+        sys.stdout.write("reached EOF")
+        sys.stdout.flush()
         os.close(ın)
         os.close(out)
     #end for
```

As you rightly note, CPython uses reference counting as its first resort for object cleanup, with its generational garbage collector as a fallback.  The problem in this case is that it seems that the `Message` object is not getting removed via normal reference counting; instead, it's persisting the lifetime of the request and ultimately left to the generational garbage collector to handle.  (This seems to hint at a reference cycle somewhere, though I haven't managed to find it yet.)

To illustrate this, I've added some GC manipulation and logging to your `fd_reap_server` sample script.  You can find the revised version (along with the client revised as detailed above) in [this Gist](https://gist.github.com/joell/8800b1a31f67a5fa6377b50ebc62a6c8).  If you run the server, and then run four sequential `fd_reap_client` processes, you should see the following output:
```
$ ./fd_reap_server
Forcing the Message object into GC generation 2.
GC counts for generations 0-2: (0, 0, 5)
Forcing the Message object into GC generation 2.
GC counts for generations 0-2: (0, 0, 6)
Forcing the Message object into GC generation 2.
GC counts for generations 0-2: (0, 0, 7)
Forcing the Message object into GC generation 2.
GC counts for generations 0-2: (0, 0, 8)
```

The GC collections of generation 0 and 1 are intended to force the `Message` object into generation 2, where it is unlikely to be collected for some time, making it easier to catch and observe what's happening.  As you can see, each subsequent client invocation increases the number of objects in generation 2 by 1, due to the dangling `Message` object from that request.

To further illustrate this, if you run `lsof` on the `fd_reap_server` process at this point, you can see the 4 dangling `dup()`'d write-end pipe file descriptors 7-10 that are still there due to the uncollected `Message` objects.  If you were to continue to send requests, you should see that collection of pipe file descriptors grow by one with each request (until Python finally triggers a generation-2 garbage collection).
```
python3 7842 joell    0u      CHR              136,1      0t0         4 /dev/pts/1
python3 7842 joell    1u      CHR              136,1      0t0         4 /dev/pts/1
python3 7842 joell    2u      CHR              136,1      0t0         4 /dev/pts/1
python3 7842 joell    3u  a_inode               0,13        0     11767 [eventpoll]
python3 7842 joell    4u     unix 0x000000001d229eb1      0t0    234682 type=STREAM
python3 7842 joell    5u     unix 0x0000000014566d21      0t0    234683 type=STREAM
python3 7842 joell    6u     unix 0x000000008dad7849      0t0    234684 type=STREAM
python3 7842 joell    7w     FIFO               0,12      0t0    237245 pipe
python3 7842 joell    8w     FIFO               0,12      0t0    232028 pipe
python3 7842 joell    9w     FIFO               0,12      0t0    240195 pipe
python3 7842 joell   10w     FIFO               0,12      0t0    228110 pipe
```

Thank you once again for such a prompt response and devoting your time and attention to this issue!  Please let me know if, perhaps for differences in our runtime environments, you still cannot reproduce this issue with the revised sample code.

**Comment by ldo** [comment]
That is correct, you have found an issue with reference circularities that I also ran into while trying to reproduce the problem you mentioned. In my case, I observed it by listing the `/proc/«server»/fd` directory, and noting several dozen pipe FDs remaining open.

I tried doing a fix or two (you will see that in the recent commit history), and you can try patching that in to see its effect. But I don’t think it’s a correct fix for the general case, which is why I have reverted it for now.

I have put on my todo list to 1) get to the bottom of the exact cause of the circularities (they seem to be happening on all 3 of the arguments to the `_message_interface_dispatch` internal routine), and 2) come up with a proper fix that won’t impact async request completions.

**Comment by ldo** [comment]
I believe I have finally tracked down and fixed the circularity issue. It happens with the definition of an inner function that makes nonlocal outer references, even if the function is never called!

Please confirm the fix works for you.

**Comment by joell** [comment]
Yes, the fix works for me!  The server process no longer accumulates file descriptors.  Thank you! :smiley:
